File: C++/3152. Special Array II.cpp
class Solution {
public:
    vector<bool> isArraySpecial(vector<int>& n, vector<vector<int>>& q) {
        vector<int>a;
        for(int i=0;i<n.size()-1;i++){
            if(n[i]%2!=n[i+1]%2)
            a.push_back(0);
            else
            a.push_back(1);
        }
        vector<int>prefix(a.size()+1,0);
        for(int i=0;i<a.size();i++)prefix[i+1]=prefix[i]+a[i];
        vector<bool>v;
        for(int i=0;i<q.size();i++){
            if(prefix[q[i][1]]-prefix[q[i][0]]>0)
            v.push_back(false);
            else
            v.push_back(true);
        }
        return v;
    }
};

File: JAVA/3152. Special Array II.java
class Solution {
    public boolean[] isArraySpecial(int[] n, int[][] q) {
        int[] a = new int[n.length - 1];
        for (int i = 0; i < n.length - 1; i++) {
            a[i] = (n[i] % 2 == n[i + 1] % 2) ? 1 : 0;
        }
        
        int[] prefix = new int[a.length + 1];
        for (int i = 0; i < a.length; i++) {
            prefix[i + 1] = prefix[i] + a[i];
        }
        
        boolean[] result = new boolean[q.length];
        for (int i = 0; i < q.length; i++) {
            if (prefix[q[i][1]] - prefix[q[i][0]] > 0) {
                result[i] = false;
            } else {
                result[i] = true;
            }
        }
        
        return result;
    }
}
